File: ios/TPStreamsRNPlayerViewManager.m
#import <React/RCTViewManager.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(TPStreamsRNPlayerViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(videoId, NSString)
RCT_EXPORT_VIEW_PROPERTY(accessToken, NSString)
RCT_EXPORT_VIEW_PROPERTY(shouldAutoPlay, BOOL)
RCT_EXPORT_VIEW_PROPERTY(startAt, double)
RCT_EXPORT_VIEW_PROPERTY(enableDownload, BOOL)
RCT_EXPORT_VIEW_PROPERTY(offlineLicenseExpireTime, double)
RCT_EXPORT_VIEW_PROPERTY(showDefaultCaptions, BOOL)
RCT_EXPORT_VIEW_PROPERTY(downloadMetadata, NSString)

// Event props
RCT_EXPORT_VIEW_PROPERTY(onCurrentPosition, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onDuration, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onIsPlaying, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPlaybackSpeed, RCTDirectEventBlock)

RCT_EXPORT_VIEW_PROPERTY(onPlayerStateChanged, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onIsPlayingChanged, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPlaybackSpeedChanged, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onIsLoadingChanged, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAccessTokenExpired, RCTDirectEventBlock)

// Player commands
RCT_EXTERN_METHOD(play:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(pause:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(seekTo:(nonnull NSNumber *)node position:(nonnull NSNumber *)position)
RCT_EXTERN_METHOD(setPlaybackSpeed:(nonnull NSNumber *)node speed:(nonnull NSNumber *)speed)
RCT_EXTERN_METHOD(getCurrentPosition:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(getDuration:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(isPlaying:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(getPlaybackSpeed:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(setNewAccessToken:(nonnull NSNumber *)node newToken:(nonnull NSString *)newToken)

@end

File: ios/TPStreamsRNPlayerViewManager.swift
import Foundation
import React
import CoreMedia
import AVFoundation

@objc(TPStreamsRNPlayerViewManager)
class TPStreamsRNPlayerViewManager: RCTViewManager {
    override func view() -> UIView! {
        return TPStreamsRNPlayerView()
    }

    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    // MARK: - Player Commands - Simply delegate to the view
    
    @objc func seekTo(_ node: NSNumber, position: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
            
            component?.seekTo(position: position.doubleValue)
        }
    }
    
    @objc func play(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
            
            component?.play()
        }
    }
    
    @objc func pause(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
            
            component?.pause()
        }
    }
    
    @objc func setPlaybackSpeed(_ node: NSNumber, speed: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
            
            component?.setPlaybackSpeed(speed.floatValue)
        }
    }
    
    // MARK: - Methods that trigger events - Simply delegate to the view
    
    @objc func getCurrentPosition(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
      
            component?.getCurrentPosition()
        }
    }
    
    @objc func getDuration(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
            
            component?.getDuration()
        }
    }
    
    @objc func isPlaying(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
            
            component?.isPlaying()
        }
    }
    
    @objc func getPlaybackSpeed(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
            
            component?.getPlaybackSpeed()
        }
    }

    @objc func setNewAccessToken(_ node: NSNumber, newToken: NSString) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as? TPStreamsRNPlayerView
            
            component?.setNewAccessToken(newToken as String)
        }
    }
}